proposals/002-Auttomaticly-reload-on-Kubernetes-config-change.md

```diff
@@ -0,0 +1,75 @@
+|      Kafka Proxy      |    aggregator    |    not aggregated     |                       Virtual Kafka Cluster CR                       |
+| Virtual Kafka Cluster | aggregator, leaf |      Kafka Proxy      | Virtual Kafka Cluster CR, Kafka Protocol Filter CR, Kafka Service CR |
+| Kafka Protocol Filter |       leaf       | Virtual Kafka Cluster |                       Kafka Protocol Filter CR                       |
+|     Kafka Service     |       leaf       | Virtual Kafka Cluster |                 Kafka Service CR, Kubernetes Secrets                 |
+|  Kafka Proxy Ingress  |       leaf       |  Kafka Proxy Cluster  |                        Kafka Proxy Ingress CR                        |
+
+To illustrate
+The Kafka Service Reconciler, KSR, will: 
+1. Watch for Kafka Service CRs to appear
+2. Parse the KafkaService
+   1. Follow any references to external secretes
+   2. For each valid secret
+      1. Compute a checksum of secret UID & Generation
+3. Annotate the KafkaService with `kroxylicious.io/dependants-checksum = <computedChecksum>`  
+
+
+[^1]: For simplicity CRC32 as that's already supported by the JDk. However, the algorithm used is immaterial.
+## Advantages
+Maintains the same logical division between reconcilers thus avoiding the need for the aggregating reconciler to watch all resources involved and thus the need for it to understand the details of the full configuration model and its kubernetes representation. 
+
+## Limitations
+1. The blast radius of any config change is still the whole proxy process. I.e. a malformed config file still prevents the whole proxy from starting.
+This is largely mitigated by the fact the config file is in the kubernetes case is programmatically generated, and we can validate that before applying it.
+2. Any change to the config model requires the entire proxy process to restart. I.e. renewing the certificate of one virtual cluster requires the all connections to the hosting proxy instance to be disconnected and reconnect.
+3. The operator requires permissions to actually get all the resources (in order to obtain the `metadata.generation` and `metadata.uid`). While it can just query for the specific metadata the Kubernetes RBAC model works at verb/resource level not to fields within resources. 
+
+# Rejected Alternatives
+1. Have the Kafka Proxy reconciler (KPR) watch all resources involved for changes and trigger on any update. This has a variety of issues
+   1. We conflate concerns between reconcilers as we end up with two (or more) reconcilers watching the same resource.
+   2. The KPR can, and likely will, be triggered on invalid or intermediate states thus making for vastly more complicated error handling and recovery
+   3. It becomes much harder for operators to work out what triggered a particular change to the running processes.
+2. The proxy process monitors the config file(s) for changes 
```

Comment:
I think *eventually* we will give the Proxy the ability to watch files containing things like TLS material and passwords and have those reloaded on the fly.   As you allude to, there's a fair amount of work in doing that.   There's the mechanism of watching the files, having to reload the contents and apply the result to the model and get things like SSL contexts to refresh (both Netty and JDK flavours).  Should we break existing TCP connections and force the clients to reconnect?   That's a lot of work.

I am trying to say, I think this proposal is right one for now.

Reply:
Yes, fully agree. The proxy should be re-loading much more automatically and in a less drop everything and restart fashion but our time is better spent else where for now.

Reply:
But I'm very glad to hear you agree!